New implementation for tree-sitter based local detection (#58924)
sourcegraph/sourcegraph: docker-images/syntax-highlighter/crates/scip-syntax/README.md

> # scip-locals

Reviewer: ```suggestion
# scip-syntax
```

## Globals (To be documented)

> ## Symbols (To be documented)

Reviewer: If this is a TODO for the future, it'd be nice to file an issue and link it here.

Author: This was more aspirational... I'm definitely not getting to it this in this PR, and I don't see a reason why this needs to be documented any more urgent than any other part of our code. I've decided to remove these headers.

## Locals

`src/locals.rs` implements an evaluator for our tree-sitter query based DSL to label local definitions and references for various programming languages using purely syntactic information. This data is used to enable fast and lightweight file-local navigation in the blob view.

Queries describing the local binding structure of various programming languages are maintained in `queries/*/scip-locals.scm`.

### The Query DSL

For specifying how to resolve local bindings we use the [tree-sitter query language] and a set of custom captures and properties. The three main concepts are _scopes_, _definitions_, and _references_.

> ### Scopes

Reviewer: nit: The heading structure is a bit off, either this should be a level 4 heading and below should be level 5, or the one above it should be level 2.

> Scopes are specified by labeling a capture as a `@scope[.kind]`. The optional scope kind can be used to hoist definitions to scopes of that kind. There is an implicit top-level scope that is of kind `"global"`

Reviewer: Please specify if `.kind` can be any string or if it needs to be one of a selected set of strings.

Reviewer: "hoisting" is defined in the scoping.md file but it is used here. It'd be nice to have a small section in the Design.md file with just `Hoisting`, give a couple of examples, and link that section's heading from here.

Author: For now scope kinds are arbitrary strings. I think we might want to move to using scip terminology for these (if that exists), and check these when reading in the locals query.

I've added a definition for hoisting in the scoping document and linked to it.

#### Examples

```scm
(block @scope)
(function_declaration @scope.function)
```

### Definitions

Definitions are specified by labeling a capture as a `@definition`.

#### Lexical scoping

```scm
(variable_definition (identifier) @definition)
```

The default behaviour for a definition is to be visible to all references that appear lexically _after_ the definition.

```js
print(my_var) // Will not be resolved

let my_var = 10
print(my_var) // Will be resolved
```

#### Hoisting

> If you want a definition to be _hoisted_ instead, you can specify what kind of scope it should be hoisted to.

Reviewer: ```suggestion
If you want a definition to be _hoisted_ to the start of a scope instead, you can specify the kind of the nearest enclosing scope it should be hoisted to.
```

```scm
(function_definition
 (identifier) @definition
 #set! "hoist" "function")
```

> It'll then be visible from the closest parent scope with the given kind. If no parent with the given kind is found, the definition will be visible in the global scope.

Reviewer: ```suggestion
The definition will be visible to the nearest enclosing scope with kind `function`. If no such enclosing scope is found, the definition will be visible in the global scope.
```

> ```js

Reviewer: This example is a bit confusing as it is only talking about the final fallback case, and not the more common case related to hoisting to the start of a scope like Python.

Author: I've made the example a little more complex to cover both cases.

> my_func(10) // Will be resolved

Reviewer: ```suggestion
my_func(10) // Will be resolved as `my_func` will be visible at global scope
```

Author: Done

function my_func(x) {
  print(x)
}
```

#### First assignment is Definition

> Certain languages (eg. Python or Matlab) do not have special syntactic forms for introducing variables.

Reviewer: ```suggestion
Certain languages (Python, MATLAB etc.) do not have special syntactic forms for introducing variables.
```

Instead the first assignment of a variable is considered to be its definition, and all further ones are references.

> To support this in our DSL you can can mark a `@definition` as a 'def_ref'.

Reviewer: Very very minor nit, but it's a little bit jarring to see some paragraphs with all sentences on one line, and some paragraphs with one sentence per line. Please stick to a single style within a file, doesn't matter which one.

(I personally prefer https://sembr.org/, but we don't use it much, so you're not required to use it.)

Author: Thank you for spotting this! I generally try to use one sentence per line, because it makes line-based diffs so much better.

Unfortunately Github doesn't render these linebreaks as spaces, but I prefer good diffs over aesthetics in this case.

```scm
(assignment
 (identifier) @definition
 #set! "def_ref")
```

If you also specify a hoist level, only existing assignments that match the current hoist-level will be considered when deciding between definition and reference.

> As an example here's how Python's hoisted ref_defs would get resolved.

Reviewer: ```suggestion
As an example, here's how local variables in Python functions could be handled.
```
